apps/nextjs/pages/login/[provider].tsx

```diff
@@ -0,0 +1,88 @@
+import { GetServerSideProps } from "next";
+import dynamic from "next/dynamic";
+import Router from "next/router";
+import { useEffect } from "react";
+
+import { ContentWrapper } from "components/Wrapper";
+import { AuthCompleteResponse } from "pages/api/auth/complete";
+import { AUTH_PROVIDERS } from "utils/constants";
+import { getJsonPostSettings } from "utils/fetch";
+import { getString } from "utils/urlQuery";
+
+interface PageProps {
+  code: string | null;
+  provider: string;
+  state: string | null;
+}
+
+function loginPage(props: PageProps): JSX.Element | null {
+  const { code, provider, state: receivedState } = props;
+  const expectedState = localStorage.getItem(`${provider}LoginState`);
+
+  if (expectedState === null) {
+    throw new Error(
+      `AuthError: state token for ${provider} not found in localStorage`
+    );
+  }
+
+  if (receivedState !== expectedState) {
+    throw new Error(
+      `AuthError: state token received from ${provider} doesn't match expected value`
+    );
+  }
+
+  if (code === null) {
+    throw new Error('AuthError: missing required parameter "code"');
+  }
+
+  useEffect(() => {
+    (async () => {
+      const res = await postNextBackend(provider, code);
+      localStorage.removeItem(`${provider}LoginState`);
+
+      if (!res.ok) {
+        throw new Error(`AuthError: ${res.error}`);
+      }
+
+      // TODO: store res.sessionId to global state.
+      Router.push("/");
+    })();
+  }, []);
+
+  return (
+    <ContentWrapper>
+      <p>Processing login request...</p>
+    </ContentWrapper>
+  );
+}
+
+// Prevent SSR since localstorage is not available on server-side.
```

Comment:
Even if we could solve this, why would we want to render the page on server, if it can't do its job (check the `state` parameter) on server-side? Then we'd just end up rendering the page twice?

Reply:
I _think_ you can pre-render the page on the server itself (if we have some "redirecting" UI) and have the redirect JS execute only on the client, which should make it slightly more responsive.

That said, it honestly is such a miniscule improvement & I'm unsure if it'd even be an improvement, we can probably leave it as is.

Reply:
Left as was.